components/support/jwcrypto/src/aes.rs

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Wrappers we use around rc_crypto's AES implementation.
+
+use crate::{
+    error::{JwCryptoError, Result},
+    CompactJwe, EncryptionAlgorithm, JweHeader,
+};
+use rc_crypto::{aead, rand};
+
+/// Does the AES-encrypt heavy-lifting for the schemes supported by this crate.
+pub(crate) fn aes_gcm_encrypt(
+    data: &[u8],
+    protected_header: JweHeader,
+    secret: &[u8],
+) -> Result<CompactJwe> {
+    let encryption_algorithm = match protected_header.enc {
+        EncryptionAlgorithm::A256GCM => &aead::AES_256_GCM,
+    };
+    let sealing_key = aead::SealingKey::new(encryption_algorithm, secret)?;
+    let additional_data = serde_json::to_string(&protected_header)?;
+    let additional_data =
+        base64::encode_config(additional_data.as_bytes(), base64::URL_SAFE_NO_PAD);
+    let additional_data = additional_data.as_bytes();
+    let aad = aead::Aad::from(additional_data);
+    let mut iv: Vec<u8> = vec![0; 12];
+    rand::fill(&mut iv)?;
+    let nonce = aead::Nonce::try_assume_unique_for_key(encryption_algorithm, &iv)?;
+    let mut encrypted = aead::seal(&sealing_key, nonce, aad, data)?;
+
+    let tag_idx = encrypted.len() - encryption_algorithm.tag_len();
+    let auth_tag = encrypted.split_off(tag_idx);
+    let ciphertext = encrypted;
+
+    CompactJwe::new(
+        Some(protected_header),
+        None,
+        Some(iv),
+        ciphertext,
+        Some(auth_tag),
+    )
+}
+
+/// Does the AES-decrypt heavy-lifting for the schemes supported by this crate
+pub(crate) fn aes_gcm_decrypt(
+    jwe: &CompactJwe,
+    encryption_algorithm: &EncryptionAlgorithm,
```

Comment:
Could we read the `enc` parameter from the protected header of `jwe` rather than requiring the caller to pass it in here? Seems like an opportunity to avoid another potential footgun for the caller, where they fail to check that `enc` has the expected value.